### PR TITLE
Make asumeRole using SAML assertion

### PR DESCRIPTION
**Pull request by benjaminrene**
* **Please check if the PR fulfills these requirements**
- Make asumeRole using SAML assertion #140 

 The commit message describes your change
- [x] Tests for the changes have been added if possible (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)
- [x] Changes are mentioned in the changelog (for bug fixes / features)

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
New features :
- Add SAML assertion authentication with a assumeRole request against AWS
- Add withRole step to switch role during the same session

### CODE REVIEW

**Comment by hoegertn** [comment]
Hi, thanks for writing this PR. I do not really understand what `withRole` is supposed to do? This is exactly the desired behavior of `withAWS`.

There is much code duplication with these both steps, too.

**Comment by benjaminrene** [comment]
Hi, thanks for reviewing. Indeed, `withRole` is pretty much the same as `withAWS` but just provide assume role. Here is my use case :

I need to authenticate against AWS with a `assumeRoleWithSAML` (with a generic 'principal' role). I cannot do any operation with this role, it's just use to assume another role which has the right to operate with my AWS resources. If we don't keep `withRole` my pipeline will look like something like this : 
```groovy
withAWS(role: 'myGenricRole', roleAccount: 'arn:...123', principalArn: 'arn:...', samlAssertion: 'base64') {
    withAWS(role: 'myRole', roleAccount: '...') {
        // do smth
    }
}
```

I found that a bit uggly so, I had the idea to add a `withRole`step :
```groovy
withAWS(role: 'myGenricRole', roleAccount: 'arn:...123', principalArn: 'arn:...', samlAssertion: 'base64') {
    withRole(role: 'myRole', roleAccount: '...') {
        // do smth
    }
}
```
*NB: I cannot authenticate with the role that can operate oon my resources*

This is why I add this step. I understand your question, and I agree that there's duplicated code. I have some issues with the shared EnvVars if I factorize the assume role behavior in the `withRole` step, but you have probably a solution ? 👍 

Regards,
Benjamin

**Comment by hoegertn** [comment]
I really see your use case, as some of my customers do the same. But I do not see a reason to have another step do the exact same thing.

The first snippet is exactly what is meant to be. You "login" the first time and then again using the obtained credentials.

**Comment by benjaminrene** [comment]
As you wish, I will remove this new step. I will update the PR this afternoon.

File: src/main/java/de/taimos/pipeline/aws/WithAWSStep.java [hunk 1]
```diff
@@ -332,46 +352,36 @@ private void withCredentials(@Nonnull Run<?, ?> run, @Nonnull EnvVars localEnv)
 
 					localEnv.override(AWSClientFactory.AWS_ACCESS_KEY_ID, awsCredentials.getAWSAccessKeyId());
 					localEnv.override(AWSClientFactory.AWS_SECRET_ACCESS_KEY, awsCredentials.getAWSSecretKey());
-
-					this.envVars.overrideAll(localEnv);
 				} else {
 					throw new RuntimeException("Cannot find a Username with password credential with the ID " + this.step.getCredentials());
 				}
+			} else if (!StringUtils.isNullOrEmpty(this.step.getSamlAssertion())) {
+				localEnv.override(AWSClientFactory.AWS_ACCESS_KEY_ID, "access_key_not_used_will_pass_through_SAML_assertion");
```
Comment (hoegertn) [inline src/main/java/de/taimos/pipeline/aws/WithAWSStep.java]: Is there a special reason for this values?

Comment (benjaminrene) [reply src/main/java/de/taimos/pipeline/aws/WithAWSStep.java]: Yes, when authenticate with assumeRoleWithSAML, the sts client will still try to use creds (even if none are needed, the AWS SDK is a bit buggy on this point). So I just set some dummy creds.

File: README.md [hunk 2]
```diff
@@ -625,6 +633,10 @@ ec2ShareAmi(
 
 ## current master
 
+## 1.34
+* add support for assume role with SAML assertion authentication (#140)
+* add withRole step to assume different role with a same AWS session
```
Comment (hoegertn) [inline README.md]: Can you remove this, please.

Comment (benjaminrene) [reply README.md]: Everything or juste the last line ?
If everything, please update the pr template, because it's ask to mention changes in the changelog.

Comment (hoegertn) [reply README.md]: Sorry for the confusion. Thank you for writing a changelog at all. I just meant the line for `withRole`.

**Comment by hoegertn** [comment]
Thanks for the contribution. I will test it and release it this week.

**Comment by jspiewak** [comment]
This change appears to now require the `region` parameter on `withAWS`, was that intended?

Without the parameter:

```
java.lang.NullPointerException
	at de.taimos.pipeline.aws.utils.IamRoleUtils.selectPartitionName(IamRoleUtils.java:37)
	at de.taimos.pipeline.aws.WithAWSStep$Execution.withRole(WithAWSStep.java:371)
	at de.taimos.pipeline.aws.WithAWSStep$Execution.start(WithAWSStep.java:292)
	at org.jenkinsci.plugins.workflow.cps.DSL.invokeStep(DSL.java:270)
```

**Comment by hoegertn** [comment]
This was not intended. Sorry for any inconvenience caused. I added a test for this and I am fixing it right now.

@benjaminrene Any reason you changes `this.step.getRegion()` to the envvars `AWS_REGION` (src/main/java/de/taimos/pipeline/aws/WithAWSStep.java:372)

**Comment by benjaminrene** [comment]
Obviously I made a mistake, that was not intended. Sorry the inconvenience.

**Pull request merged**